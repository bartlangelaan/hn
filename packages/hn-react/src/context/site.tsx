import { Site } from 'hn';
import * as React from 'react';
import globalSite from '../utils/site';

// Create a new context.
const HnContext = React.createContext(globalSite);
const { Consumer, Provider } = HnContext;

// The site provider is the same as the 'Provider', but changes 'site' to
// 'value'.
export const SiteProvider = ({
  site,
  children,
}: {
  site: Site;
  children: React.ReactNode;
}) => <Provider value={site} children={children} />;

// The consumer is exported as both Site and SiteConsumer. The SiteConsumer can
// be used if also the Site from 'hn' is imported, so the names don't overlap.
export { Consumer as Site, Consumer as SiteConsumer };

// There is also a HOC withSite that adds the site to the props of the
// component. Source of the HOC for Typescript:
// https://stackoverflow.com/a/50613946/1907875
export interface InjectedSiteProps {
  site: Site;
}

export function withSite<P extends InjectedSiteProps>(
  Component: React.ComponentType<P>,
) {
  const WrappedComponent: any = Component;
  return function ComponentWithSite(
    props: Pick<P, Exclude<keyof P, keyof InjectedSiteProps>>,
  ) {
    return (
      <Consumer>{site => <WrappedComponent {...props} site={site} />}</Consumer>
    );
  };
}

export function useHnSite(): Site {
  // @ts-ignore
  return React.useContext(HnContext);
}
