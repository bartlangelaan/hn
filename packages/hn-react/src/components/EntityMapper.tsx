import { Site } from 'hn';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { SiteConsumer } from '../context/site';

const getNested = require('get-nested'); // tslint:disable-line:no-var-requires

const entityCache = new Map<() => ImportReturn, React.ReactType>();
export const clearEntityCache = () => entityCache.clear();

type EntityMapperInnerProps = EntityMapperProps & { site: Site };

export class EntityMapper extends React.Component<
  EntityMapperInnerProps,
  { error: Error | null }
> {
  state = {
    error: null,
  };

  lastAssuredProps: EntityMapperInnerProps | null = null;

  /**
   * If this component exists in a tree that is invoked with the waitForHnData function, this function is invoked.
   * Only after the promise is resolved, the component will be mounted. To keep the data fetched here, we assign the
   * state to the hnContext provided by the DrupalPageContextProvider. This way, the state will be preserved trough
   * multiple renders.
   */
  async bootstrap() {
    // Make sure that if this is an async component, we have it saved in the entity cache.
    return assureComponent(this.props);
  }

  lastRequest: symbol | undefined;

  componentDidMount() {
    this.componentDidUpdate();
  }

  componentDidUpdate() {
    /**
     * This function is almost the same as the componentDidUpdate in the DrupalPage.
     * Please check there for more comments, context and reference.
     */
    const lastRequest = (this.lastRequest = Symbol());

    if (this.isReady()) {
      this.lastAssuredProps = this.props;
      return;
    }

    (async () => {
      await assureComponent(this.props);

      if (lastRequest !== this.lastRequest) {
        return;
      }

      // istanbul ignore next
      if (!this.isReady()) {
        throw new Error('The component cannot be loaded.');
      }

      this.lastAssuredProps = this.props;

      this.forceUpdate();
    })().catch(error => {
      this.setState({ error });
    });
  }

  componentWillUnmount() {
    this.lastRequest = undefined;
  }

  isReady(props = this.props) {
    return typeof getComponentFromMapper(props) !== 'undefined';
  }

  render() {
    /**
     * This render function is also almost the same technique as the DrupalPage render.
     * Please check that function for comments.
     */
    if (this.state.error) {
      throw this.state.error;
    }

    const canRenderFromProps = this.isReady();

    const props = canRenderFromProps ? this.props : this.lastAssuredProps;

    if (!props) {
      return null;
    }

    const EntityComponent = getComponentFromMapper(props);

    if (EntityComponent === null) {
      return null;
    }

    // istanbul ignore next
    if (!EntityComponent) {
      throw new Error(
        "We received data that we can't render. This shouldn't be possible.",
      );
    }

    const data = props.site.getData(props.uuid);

    return (
      <EntityComponent
        bundle={getBundle(data)}
        paragraph={data}
        entity={data}
        {...props.entityProps}
      />
    );
  }

  static propTypes = {
    asyncMapper: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.oneOfType([PropTypes.shape({}), PropTypes.func]),
    ]),
    entityProps: PropTypes.shape({}),
    mapper: PropTypes.oneOfType([
      PropTypes.shape({}),
      PropTypes.func,
      PropTypes.bool,
    ]),
    uuid: PropTypes.string.isRequired,
  };

  static defaultProps = {
    asyncMapper: false,
    entityProps: {},
    mapper: false,
  };
}

export type ImportReturnResolved =
  | React.ReactType
  | { default: React.ReactType };
export type ImportReturn = Promise<ImportReturnResolved>;

export interface ObjectMapper {
  [uuid: string]: React.ReactType;
}
export interface ObjectMapperAsync {
  [uuid: string]: () => ImportReturn;
}

export type functionMapper = (
  entity: object,
  bundle: string,
) => React.ReactType;
export type functionMapperAsync = (
  entity: object,
  bundle: string,
) => () => ImportReturn;

export type mapperType =
  | ObjectMapper
  | ObjectMapperAsync
  | functionMapper
  | functionMapperAsync;

export interface EntityMapperPropsBase {
  entityProps?: object;
  uuid: string;
}

export interface EntityMapperPropsMapperAsync {
  mapper?: undefined;
  asyncMapper: ObjectMapperAsync | functionMapperAsync;
}

export interface EntityMapperPropsMapperAsyncClassic {
  mapper: ObjectMapperAsync | functionMapperAsync;
  asyncMapper: true;
}

export interface EntityMapperPropsMapperSync {
  mapper: ObjectMapper | functionMapper;
  asyncMapper?: false;
}

export type EntityMapperPropsMapper =
  | EntityMapperPropsMapperAsync
  | EntityMapperPropsMapperAsyncClassic
  | EntityMapperPropsMapperSync;

export type EntityMapperProps = EntityMapperPropsBase & EntityMapperPropsMapper;

const EntityMapperWrapper = React.forwardRef<EntityMapper, EntityMapperProps>(
  (props, ref) => (
    <SiteConsumer>
      {site => <EntityMapper {...props} site={site} ref={ref} />}
    </SiteConsumer>
  ),
);

export default EntityMapperWrapper;

/**
 * This gets a component from some mapper data.
 * This can also be a promise resolving to a component.
 */
function getComponentOrPromiseFromMapper(
  props: EntityMapperInnerProps,
):
  | { type: 'component'; value: undefined | React.ReactType }
  | { type: 'promise'; value: undefined | (() => ImportReturn) } {
  // This gets the entity from the site, based on the uuid.
  const entity = props.site.getData(props.uuid);

  // This should give back a bundle string, that is used in the mapper.
  const bundle = getBundle(entity);

  // If it's not an async mapper, we can just use the 'mapper' prop.
  if (!props.asyncMapper) {
    return {
      type: 'component',
      value:
        typeof props.mapper === 'function'
          ? props.mapper(entity, bundle)
          : props.mapper[bundle],
    };
  }

  // We check which style of the mapper it is.
  const entityComponentMapper =
    props.asyncMapper === true ? props.mapper : props.asyncMapper;
  const entityComponentResult =
    typeof entityComponentMapper === 'function'
      ? entityComponentMapper(entity, bundle)
      : entityComponentMapper[bundle];

  return {
    type: 'promise',
    value: entityComponentResult,
  };
}

/**
 * This makes sure the data for this url is ready to be rendered.
 */
export async function assureComponent(props: EntityMapperInnerProps) {
  const component = getComponentOrPromiseFromMapper(props);

  // If it's not a promise, return.
  // Nothing to assure.
  if (component.type === 'component') {
    return;
  }

  // If no value was returned from the mapper, we can't assure the component.
  if (!component.value) {
    return;
  }

  // Check if we already cached this component.
  if (entityCache.has(component.value)) {
    return;
  }

  // If not, we get this entity now.
  const entityComponentResult = await component.value();

  // We save the entity in the cache.
  if (
    typeof entityComponentResult === 'object' &&
    typeof entityComponentResult.default !== 'undefined'
  ) {
    entityCache.set(component.value, entityComponentResult.default);
  } else {
    entityCache.set(component.value, entityComponentResult as React.ReactType);
  }
}

/**
 * Returns the component if it is availible right now.
 * Returns null if the component isn't available, and never will be.
 * Returns undefined if the component isn't available, but maybe will be in the future.
 */
export function getComponentFromMapper(
  props: EntityMapperInnerProps,
): React.ReactType | null | undefined {
  const entityComponent = getComponentOrPromiseFromMapper(props);

  if (
    entityComponent.type === 'promise' &&
    typeof entityComponent.value !== 'undefined'
  ) {
    return entityCache.get(entityComponent.value);
  }
  if (entityComponent.type === 'component') {
    return entityComponent.value || null;
  }
  return null;
}

function getBundle(entity) {
  return getNested(
    () => `${entity.__hn.entity.type}__${entity.__hn.entity.bundle}`,
    '_fallback',
  );
}
