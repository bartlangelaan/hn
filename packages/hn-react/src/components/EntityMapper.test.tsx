import * as React from 'react';
import * as renderer from 'react-test-renderer';
import site from '../utils/site';
import {
  asyncMapper,
  entity,
  entity2,
  mapper,
  uuid,
  uuid2,
} from '../utils/tests';
import waitForHnData from '../utils/waitForHnData';
import EntityMapper, {
  EntityMapper as InnerEntityMapper,
  clearEntityCache,
  assureComponent,
} from './EntityMapper';

jest.mock('../utils/site', () => {
  return require('../utils/tests').mockSite();
});

jest.mock('util-deprecate', () => jest.fn(func => func));

describe('EntityMapper', async () => {
  afterEach(() => {
    clearEntityCache();
  });

  test('with required props', async () => {
    const component = <EntityMapper uuid={uuid} mapper={mapper} />;

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('non-existing entity', async () => {
    const component = (
      <EntityMapper uuid={'non-existing-entity'} mapper={mapper} />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with required props & asyncMapper', async () => {
    const component = <EntityMapper uuid={uuid} asyncMapper={asyncMapper} />;

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with all props', async () => {
    const component = (
      <EntityMapper
        entityProps={{ testEntityProp: true }}
        uuid={uuid}
        mapper={mapper}
      />
    );

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('with all props & asyncMapper', async () => {
    const component = (
      <EntityMapper
        entityProps={{ testEntityProp: true }}
        uuid={uuid}
        asyncMapper={asyncMapper}
      />
    );

    expect(renderer.create(component).toJSON()).toBe(null);

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('asyncMapper as boolean (deprecated)', async () => {
    const component = (
      <EntityMapper uuid={uuid} mapper={asyncMapper} asyncMapper />
    );

    expect(renderer.create(component).toJSON()).toBe(null);

    expect(
      renderer.create(await waitForHnData(component)).toJSON(),
    ).toMatchSnapshot();
  });

  test('function mapper', async () => {
    const customMapper = jest.fn((_, typeBundle) => mapper[typeBundle]);

    const component = <EntityMapper uuid={uuid} mapper={customMapper} />;

    expect(renderer.create(component).toJSON()).toMatchSnapshot();

    expect(customMapper).toHaveBeenCalledWith(
      entity,
      'unique_type_1__unique_bundle_1',
    );
  });

  test('async mapper that fails loading', async () => {
    const importLike = jest.fn(async () => {
      throw new Error('No connection.');
    });
    const customMapper = {
      unique_type_1__unique_bundle_1: importLike,
    };
    const log = jest.fn();

    class ErrorBoundary extends React.Component {
      componentDidCatch() {
        log();
      }
      render() {
        return this.props.children;
      }
    }

    // Render the EntityMapper that will fail.
    const rendererEntry = renderer.create(
      <ErrorBoundary>
        <EntityMapper uuid={uuid} asyncMapper={customMapper} />
      </ErrorBoundary>,
    );

    // When the component isn't loaded yet, should just render null.
    expect(log).toBeCalledTimes(0);
    expect(rendererEntry.toJSON()).toBe(null);

    // Let the promise reject.
    await new Promise(resolve => process.nextTick(resolve));

    // The error boundry should be called, and still null should be rendered.
    expect(log).toBeCalledTimes(1);
    expect(rendererEntry.toJSON()).toBe(null);

    // When something changes, it should retry.
    rendererEntry.update(
      <ErrorBoundary>
        <EntityMapper uuid={uuid} asyncMapper={customMapper} />
      </ErrorBoundary>,
    );
    expect(log).toBeCalledTimes(1);

    await new Promise(resolve => process.nextTick(resolve));
    expect(log).toBeCalledTimes(2);
  });

  test('assure component with mapper not supporting bundle', async () => {
    expect(assureComponent({ site, uuid, asyncMapper: {} })).resolves.toBe(
      undefined,
    );
  });

  async function setupPropsChange() {
    let resolveEntity1Promise;
    let resolveEntity2Promise;
    const customAsyncMapper = {
      unique_type_1__unique_bundle_1: jest.fn(
        () => new Promise(r => (resolveEntity1Promise = r)),
      ),
      unique_type_2__unique_bundle_2: jest.fn(
        () => new Promise(r => (resolveEntity2Promise = r)),
      ),
    };

    const rendererEntry = renderer.create(
      <EntityMapper uuid={uuid} asyncMapper={customAsyncMapper} />,
    );

    await new Promise(resolve => process.nextTick(resolve));
    expect(customAsyncMapper.unique_type_1__unique_bundle_1).toBeCalledTimes(1);

    rendererEntry.update(
      <EntityMapper uuid={uuid2} asyncMapper={customAsyncMapper} />,
    );

    await new Promise(resolve => process.nextTick(resolve));
    expect(customAsyncMapper.unique_type_2__unique_bundle_2).toBeCalledTimes(1);
    return { resolveEntity1Promise, resolveEntity2Promise, rendererEntry };
  }

  test(`change props while already loading 1`, async () => {
    const {
      resolveEntity1Promise,
      resolveEntity2Promise,
      rendererEntry,
    } = await setupPropsChange();

    resolveEntity1Promise('section');
    expect(rendererEntry.toJSON()).toBe(null);

    resolveEntity2Promise('div');
    await new Promise(resolve => process.nextTick(resolve));
    expect(rendererEntry.root.findByType('div').props.bundle).toBe(
      'unique_type_2__unique_bundle_2',
    );
  });

  test(`change props while already loading 2`, async () => {
    const {
      resolveEntity1Promise,
      resolveEntity2Promise,
      rendererEntry,
    } = await setupPropsChange();

    resolveEntity2Promise('section');
    await new Promise(resolve => process.nextTick(resolve));
    const json = rendererEntry.toJSON();
    expect(rendererEntry.root.findByType('section').props.bundle).toEqual(
      'unique_type_2__unique_bundle_2',
    );

    resolveEntity1Promise('div');
    expect(rendererEntry.toJSON()).toEqual(json);
  });

  test('changing props', async () => {
    const ref = React.createRef<InnerEntityMapper>();
    // First, render uuid1.
    const rendererEntry = renderer.create(
      <EntityMapper uuid={uuid} mapper={mapper} ref={ref} />,
    );
    // Then, change uuid1 to uuid2.
    rendererEntry.update(
      <EntityMapper uuid={uuid2} mapper={mapper} ref={ref} />,
    );
    // Wait one tick to let React update.
    await new Promise(resolve => process.nextTick(resolve));
    // This should render the entity of uuid2.
    const entity2Result = rendererEntry.toJSON();
    expect(entity2Result).toMatchSnapshot();

    expect(ref.current.isReady()).toBe(true);

    let resolveEntity1Promise;
    let resolveEntity2Promise;

    const entity1Promise: Promise<string> = new Promise(
      resolve => (resolveEntity1Promise = resolve),
    );
    const entity2Promise: Promise<string> = new Promise(
      resolve => (resolveEntity2Promise = resolve),
    );

    const customAsyncMapper = {
      unique_type_1__unique_bundle_1: () => entity1Promise,
      unique_type_2__unique_bundle_2: () => entity2Promise,
    };

    // Now we're changing the mapper for a asyncMapper, but keeping the uuid.
    rendererEntry.update(
      <EntityMapper
        uuid={uuid2}
        mapper={customAsyncMapper}
        asyncMapper
        ref={ref}
      />,
    );

    // Wait one tick to let React update.
    await new Promise(resolve => process.nextTick(resolve));

    // The EntityMapper shouldn't have changed, and marked as not-ready.
    expect(ref.current.isReady()).toBe(false);
    expect(rendererEntry.toJSON()).toEqual(entity2Result);

    // Now we resolve the promise of the async entity.
    resolveEntity2Promise('section');
    await new Promise(resolve => process.nextTick(resolve));

    // It should now be ready, and its contents should have changed.
    expect(ref.current.isReady()).toBe(true);
    const asyncEntity2Result = rendererEntry.toJSON();
    expect(
      ((rendererEntry.root.children[0] as renderer.ReactTestInstance)
        .children[0] as renderer.ReactTestInstance).type,
    ).toBe('section');

    // Now we change the uuid and add entityProps.
    rendererEntry.update(
      <EntityMapper
        uuid={uuid}
        mapper={customAsyncMapper}
        asyncMapper
        ref={ref}
        entityProps={{ testEntityProp: 'test123' }}
      />,
    );
    await new Promise(resolve =>
      process.nextTick(() => process.nextTick(resolve)),
    );

    // The EntityMapper shouldn't have changed, and marked as not-ready.
    expect(ref.current.isReady()).toBe(false);
    expect(rendererEntry.toJSON()).toEqual(asyncEntity2Result);

    // Now we resolve the promise of the async entity.
    resolveEntity1Promise('p');
    await new Promise(resolve => process.nextTick(resolve));
    expect(ref.current.isReady()).toBe(true);
    const innerwrapper = rendererEntry.root
      .children[0] as renderer.ReactTestInstance;
    const result = innerwrapper.children[0] as renderer.ReactTestInstance;
    expect(result.type).toBe('p');
    expect(result.props.testEntityProp).toBe('test123');

    // Now change the uuid prop to a non-existing uuid.
    rendererEntry.update(
      <EntityMapper
        uuid={'doesntexist'}
        mapper={customAsyncMapper}
        asyncMapper
        ref={ref}
        entityProps={{ testEntityProp: 'test123' }}
      />,
    );
    await new Promise(resolve => process.nextTick(resolve));

    expect(rendererEntry.toJSON()).toBe(null);
  });
});
