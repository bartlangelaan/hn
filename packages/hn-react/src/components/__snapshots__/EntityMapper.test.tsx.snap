// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityMapper asyncMapper as boolean (deprecated) 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;

exports[`EntityMapper changing props 1`] = `
<div
  bundle="unique_type_2__unique_bundle_2"
  className="MyCustomMapperComponentWithHOC"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_2",
          "type": "unique_type_2",
        },
      },
    }
  }
  hocProps={true}
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_2",
          "type": "unique_type_2",
        },
      },
    }
  }
/>
`;

exports[`EntityMapper function mapper 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;

exports[`EntityMapper non-existing entity 1`] = `null`;

exports[`EntityMapper non-existing entity 2`] = `null`;

exports[`EntityMapper with all props & asyncMapper 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testEntityProp={true}
/>
`;

exports[`EntityMapper with all props 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testEntityProp={true}
/>
`;

exports[`EntityMapper with all props 2`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  testEntityProp={true}
/>
`;

exports[`EntityMapper with required props & asyncMapper 1`] = `null`;

exports[`EntityMapper with required props & asyncMapper 2`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;

exports[`EntityMapper with required props 1`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;

exports[`EntityMapper with required props 2`] = `
<div
  bundle="unique_type_1__unique_bundle_1"
  className="MyCustomMapperComponent"
  entity={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
  paragraph={
    Object {
      "__hn": Object {
        "entity": Object {
          "bundle": "unique_bundle_1",
          "type": "unique_type_1",
        },
      },
    }
  }
/>
`;
